pom.xml

```diff
@@ -180,7 +167,7 @@
                 <artifactId>maven-shade-plugin</artifactId>
                 <configuration>
                     <!--
-                        embed all jackrabbit artifacts to prevent classpath clashes
+                        embed all jackrabbit artifacts, to prevent classpath clashes
```

Comment:
Usually the Oak classes are twice in the Maven test class path. Once transitively from https://github.com/apache/sling-org-apache-sling-testing-sling-mock/blob/4176be162961185bc8a08d7ef56d60c2ba8c3028/core/pom.xml#L170 and once from this shading. Therefore the order matters here. Can we somehow do an in advance check to make sure that right version is loaded?

Reply:
I implemented such a check in https://github.com/apache/sling-org-apache-sling-testing-sling-mock-oak/pull/4/commits/b1824a4fee9ad11c5d33418fc596676eac3ada9b

Reply:
when using in combination with wcm.io aem-dependencies or aem-cloud-depencies the oak.version does not matter - the version for oak-jcr from the aem-dependencies always "wins" over the transitive dependency, regardless in which order this is put here. that means this artifact can only be used with AEMaaCS, not any AEM 6.5 dependencies (even AEM 6.5.11 is only at 1.22.9, not even at 1.26 which you marked as minimum for java 14+). that's probably ok, but we have to be aware of it that lot of users will not be able to use it, esp. not if the "by default" just want to update to latest version of this test dependency.

i'm not sure if such a hard check for matching version is the best solution? probably better check for same or newer? it might be ok if sling-mock-oak is e.g. at oak 1.40 and oak-jcr from a future aem-cloud-dependencies is at 1.42 or later?

Reply:
Ok, just checking for a downgrade is probably enough, as that will most probably lead to issues (and is not deliberate). Regarding Jackrabbit-API I now shaded that as well, any version included before will still win, but I don't think its version should be managed (as the version is coupled to the Oak impl).

Reply:
ideally, the following test case should work with this new version as well:

- checkout https://github.com/wcm-io/wcm-io-testing (this is based on AEM 6.4.0 dependencies)
- in aem-mock/parent/pom.xml switch sling-mock.oak.version to 3.0.1-1.40.0-SNAPSHOT
- run the tests for /aem-mock

Reply:
I tested with ACS AEM Commons depending on wcm.io dependencies for AEM 6.4 (and it works fine): https://github.com/Adobe-Consulting-Services/acs-aem-commons/pull/2700/commits/7ed39a321c3d342612961595920beb3887d85c8d

Reply:
Building aem-mock.core fails with
```
[main] ERROR io.wcm.testing.mock.aem.context.ContextResourceResolverFactory - Failed initializing resource resolver factory, bundleContext=org.apache.sling.testing.mock.osgi.MockBundleContext@1704f67f
java.lang.RuntimeException: Unable to invoke method 'activate' for class org.apache.sling.testing.mock.sling.oak.OakMockSlingRepository
	at org.apache.sling.testing.mock.osgi.OsgiServiceUtil.invokeMethod(OsgiServiceUtil.java:347)
	at org.apache.sling.testing.mock.osgi.OsgiServiceUtil.invokeLifecycleMethod(OsgiServiceUtil.java:163)
	at org.apache.sling.testing.mock.osgi.OsgiServiceUtil.activateDeactivate(OsgiServiceUtil.java:98)
	at org.apache.sling.testing.mock.osgi.MockOsgi.registerInjectActivateService(MockOsgi.java:230)
	at org.apache.sling.testing.mock.osgi.MockOsgi.registerInjectActivateService(MockOsgi.java:247)
	at org.apache.sling.testing.mock.sling.ResourceResolverFactoryInitializer.registerServiceIfNotPresent(ResourceResolverFactoryInitializer.java:197)
	at org.apache.sling.testing.mock.sling.ResourceResolverFactoryInitializer.registerServiceIfNotPresent(ResourceResolverFactoryInitializer.java:183)
	at org.apache.sling.testing.mock.sling.ResourceResolverFactoryInitializer.setUp(ResourceResolverFactoryInitializer.java:69)
	at org.apache.sling.testing.mock.sling.MockSling.newResourceResolverFactory(MockSling.java:87)
	at io.wcm.testing.mock.aem.context.ContextResourceResolverFactory.get(ContextResourceResolverFactory.java:55)
	at io.wcm.testing.mock.aem.context.AemContextImpl.newResourceResolverFactory(AemContextImpl.java:99)
	at org.apache.sling.testing.mock.sling.context.SlingContextImpl.resourceResolverFactory(SlingContextImpl.java:151)
	at org.apache.sling.testing.mock.sling.context.SlingContextImpl.setUp(SlingContextImpl.java:135)
	at io.wcm.testing.mock.aem.context.AemContextImpl.setUp(AemContextImpl.java:104)
	at io.wcm.testing.mock.aem.junit.AemContext.access$400(AemContext.java:56)
	at io.wcm.testing.mock.aem.junit.AemContext$2.execute(AemContext.java:208)
	at io.wcm.testing.mock.aem.junit.AemContext$2.execute(AemContext.java:203)
	at io.wcm.testing.junit.rules.parameterized.RepeatedStatement.evaluate(RepeatedStatement.java:55)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.lang.IllegalStateException: Unexpected (too old) Oak version 1.8.2, expected 1.40.0. Make sure to load the `org.apache.sling.testing.sling-mock-oak` dependency before any other dependency which may transitively depend on Apache Jackrabbit Oak as well!
	at org.apache.sling.testing.mock.sling.oak.OakMockSlingRepository.activate(OakMockSlingRepository.java:89)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.sling.testing.mock.osgi.OsgiServiceUtil.invokeMethod(OsgiServiceUtil.java:338)
	... 37 more
```

This is IMHO expected as the order of dependencies prefers the old 1.8.2 version due to the transitive dependencies of https://github.com/wcm-io/wcm-io-testing/blob/679d511e90214b4fa8a5a9d3d47a3f342461a1f8/aem-mock/core/pom.xml#L109 (which is loaded prior to sling-mock-oak). I doubt that AEM mock really depends on Oak implementation classes and probably the dependency to oak-jcr should be reworked (with excludes) or alternatively put towards the end of the dependencies.

Reply:
After removing the dependency in https://github.com/wcm-io/wcm-io-testing/blob/679d511e90214b4fa8a5a9d3d47a3f342461a1f8/aem-mock/core/pom.xml#L107-L111 everything `mvn clean install` works fine on wcm-io-testing (all modules) tested against this sling-mock-oak dependency.

Reply:
But sling-mock already depends on oak-jcr, so it is always in the class path

Reply:
thanks, did not found any issue with it